### ISSUE TITLE
Implement Optimistic Concurrency Control for Database Access

### ISSUE THREAD

**Issue by NADOOITChristophBa** [NADOOITChristophBa/AI-Hub#9]
### Issue Description
Currently, our application uses a locking mechanism at the database connection level to manage concurrent access to the SQLite database. While this approach ensures safety and simplicity, it might not be the most efficient in terms of performance, especially under high concurrency scenarios.

### Proposed Change
Implement Optimistic Concurrency Control (OCC) as an alternative to our current locking mechanism. OCC allows multiple transactions to proceed in parallel, checking for data conflicts at the time of committing. This could significantly improve performance by reducing lock contention and allowing higher concurrency.

### Benefits
- **Increased Performance:** More transactions can be handled in parallel, especially in read-heavy scenarios.
- **Reduced Lock Contention:** By removing the need for exclusive locks during the entire transaction, we reduce the risk of bottlenecks.
- **Scalability:** OCC is generally more scalable, a crucial aspect as our application grows.

### Potential Challenges
- **Implementation Complexity:** OCC requires careful implementation to handle conflict detection and resolution.
- **Handling Conflicts:** We need a strategy to handle conflicts when they are detected, either by retrying the transaction or aborting it.
- **Testing:** Rigorous testing is required to ensure data integrity and consistency are maintained.

### Suggested Implementation Steps
1. Research and design an OCC approach suitable for our use case.
2. Implement a prototype on a separate branch for testing.
3. Test rigorously with different scenarios to ensure data integrity and performance improvement.
4. Gather feedback and iterate on the implementation.
5. Merge into the main branch after thorough testing and validation.

### Additional Notes
This is a significant change and should be approached in phases, with continuous testing and monitoring at each stage to ensure system stability.

---

### Labels
- Enhancement
- Performance
- Database

---
